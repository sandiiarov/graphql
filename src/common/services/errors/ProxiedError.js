// @flow

type ProxiedErrorValues = {|
  +statusCode: string,
  +url: string,
|};

/**
 * This exception is thrown if origin server does NOT respond with HTTP status code 200.
 */
class ProxiedError extends Error {
  extensions: {|
    +_proxy: ProxiedErrorValues,
    +extensions: {|
      +proxy: ProxiedErrorValues,
    |},
  |};

  constructor(message: string, originStatusCode: string, originUrl: string) {
    super(message);
    this.extensions = {
      _proxy: {
        // note: this is deprecated and will be removed
        statusCode: originStatusCode,
        url: originUrl,
      },
      extensions: {
        proxy: {
          statusCode: String(originStatusCode),
          url: originUrl,
        },
      },
    };
  }
}

export { ProxiedError };
