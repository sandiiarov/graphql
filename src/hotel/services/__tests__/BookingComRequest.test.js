// @flow

import { get } from '../BookingComRequest';
import { ProxiedError } from '../../../common/services/errors/ProxiedError';

jest.mock('node-fetch');

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  process.env.NODE_ENV = 'production';
  process.env.BOOKING_COM_TOKEN = 'booking_com_token';
});

afterEach(() => {
  jest.resetAllMocks();
  process.env.NODE_ENV = 'test';
  delete process.env.BOOKING_COM_TOKEN;
});

describe('GET request', () => {
  it('should throw exception in test environment', async () => {
    process.env.NODE_ENV = 'test';
    await expect(get('https://path/to/api')).rejects.toBeError(
      'HttpRequest should never be called in test environment. Have you forgotten to mock "https://path/to/api" with fake data response?',
    );
  });
});

describe('GET request in production', () => {
  it('sets authorization header', async () => {
    expect(await get('https://path/to/api')).toMatchSnapshot();
  });

  it('throws exception during invalid return status code', async () => {
    expect.assertions(3);

    try {
      await get('https://path/to/api?status=500');
    } catch (error) {
      expect(error).toBeInstanceOf(ProxiedError);
      expect(error.message).toBe('Status Text');
      expect(error.extensions).toMatchSnapshot();
    }
  });
});
