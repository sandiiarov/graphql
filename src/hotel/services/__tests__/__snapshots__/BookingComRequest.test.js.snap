// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET request in production sets authorization header 1`] = `
Object {
  "hint": "this is a mocked response",
  "options": Object {
    "headers": Object {
      "authorization": "booking_com_token",
    },
    "method": "GET",
  },
  "url": "https://path/to/api",
}
`;

exports[`GET request in production throws exception during invalid return status code 1`] = `
Object {
  "_proxy": Object {
    "statusCode": "500",
    "url": "https://path/to/api?status=500",
  },
  "extensions": Object {
    "proxy": Object {
      "statusCode": "500",
      "url": "https://path/to/api?status=500",
    },
  },
}
`;
